components/iis/component.go

```diff
@@ -0,0 +1,148 @@
+//	iisServer, err := iis.NewServer(ctx, awsEnv.CommonEnvironment, vm,
+//		iis.WithSite(iis.SiteDefinition{
+//			Name:            "TestSite1",
+//			BindingPort:     "*:8081:",
+//			SourceAssetsDir: srcDir,
+//		}),
+//	)
+//	if err != nil {
+//		return err
+//	}
+//	err = iisServer.Export(ctx, env.IISServer)
+func NewServer(ctx *pulumi.Context, e *config.CommonEnvironment, host *remote.Host, options ...Option) (*Component, error) {
+	if host.OS.Descriptor().Family() != os.WindowsFamily {
+		// Print flavor in case OS family don't match, as it's more precise than the family (and the .String() conversion already exists).
+		return nil, fmt.Errorf("wrong Operating System family, expected Windows, got %s", host.OS.Descriptor().Flavor.String())
+	}
+
+	params, err := common.ApplyOption(&Configuration{}, options)
+	if err != nil {
+		return nil, err
+	}
+
+	iisComponent, err := components.NewComponent(*e, host.Name(), func(comp *Component) error {
+		comp.namer = e.CommonNamer.WithPrefix(comp.Name())
+		comp.host = host
+
+		installIISCommand, err := host.OS.Runner().Command(comp.namer.ResourceName("install-iis"), &command.Args{
+			Create: pulumi.String(`
+function ExitWithCode($exitcode) {
+	$host.SetShouldExit($exitcode)
+	exit $exitcode
+}
+$result = Install-Windowsfeature -name Web-Server -IncludeManagementTools
+if (! $result.Success ) {
+	exit -1
+}
+if ($result.RestartNeeded -eq "Yes") {
+	ExitWithCode(3010)
+}
+`),
+		}, pulumi.Parent(comp))
+		if err != nil {
+			return err
+		}
+
+		timeProvider, err := time.NewProvider(ctx, comp.namer.ResourceName("time-provider"), &time.ProviderArgs{}, pulumi.DeletedWith(host))
+		if err != nil {
+			return err
+		}
+
+		waitForRebootCmd, err := time.NewSleep(ctx, comp.namer.ResourceName("wait-for-host-to-reboot"), &time.SleepArgs{
+			CreateDuration: pulumi.String("30s"),
+		}, pulumi.Provider(timeProvider), pulumi.DependsOn([]pulumi.Resource{
+			installIISCommand,
+		}))
+		if err != nil {
+			return err
+		}
+
+		ensureIISStarted, err := host.OS.Runner().Command(comp.namer.ResourceName("ensure-iis-started"), &command.Args{
+			Create: pulumi.String(`(Get-Service W3SVC).WaitForStatus('Running', '00:01:00')`),
+		}, pulumi.DependsOn([]pulumi.Resource{
+			waitForRebootCmd,
+		}))
+		if err != nil {
+			return err
+		}
+
+		if len(params.Sites) > 0 {
+			for _, site := range params.Sites {
+				dependencies := []pulumi.ResourceOption{
+					utils.PulumiDependsOn(ensureIISStarted),
+				}
+
+				if site.TargetAssetsDir == "" {
+					site.TargetAssetsDir = fmt.Sprintf("C:\\inetpub\\%s", site.Name)
+				}
+
+				if site.SourceAssetsDir != "" {
+					copyAssets, err := host.OS.FileManager().CopyAbsoluteFolder(site.SourceAssetsDir, site.TargetAssetsDir, dependencies...)
+					if err != nil {
+						return err
+					}
+					dependencies = append(dependencies, pulumi.DependsOn(copyAssets))
+				}
+
+				_, err = host.OS.Runner().Command(comp.namer.ResourceName("create-iis-site", site.Name), &command.Args{
```

Comment:
<!-- Suggestions propose improvements to the current subject. It’s important to be explicit and clear on what is being suggested and why it is an improvement. -->
**💬 suggestion**
Same here, you might want to ensure names are all differents or pulumi will fail with cryptic "not unique resource"